internal/controllers/utils/conversion_test.go

```diff
@@ -0,0 +1,143 @@
+)
+
+func TestUpdateLoadBalancerIngress(t *testing.T) {
+	const (
+		oldIP       = "1.2.3.4"
+		oldHostname = "example"
+	)
+
+	oldIngress := func() []client.Object {
+		return []client.Object{
+			&netv1.Ingress{
+				Status: netv1.IngressStatus{
+					LoadBalancer: netv1.IngressLoadBalancerStatus{
+						Ingress: []netv1.IngressLoadBalancerIngress{
+							{
+								IP:       oldIP,
+								Hostname: oldHostname,
+							},
+						},
+					},
+				},
+			},
+			&extv1beta1.Ingress{
+				Status: extv1beta1.IngressStatus{
+					LoadBalancer: extv1beta1.IngressLoadBalancerStatus{
+						Ingress: []extv1beta1.IngressLoadBalancerIngress{
+							{
+								IP:       oldIP,
+								Hostname: oldHostname,
+							},
+						},
+					},
+				},
+			},
+			&netv1beta1.Ingress{
+				Status: netv1beta1.IngressStatus{
+					LoadBalancer: netv1beta1.IngressLoadBalancerStatus{
+						Ingress: []netv1beta1.IngressLoadBalancerIngress{
+							{
+								IP:       oldIP,
+								Hostname: oldHostname,
+							},
+						},
+					},
+				},
+			},
+			&kongv1beta1.TCPIngress{
+				Status: kongv1beta1.TCPIngressStatus{
+					LoadBalancer: corev1.LoadBalancerStatus{
+						Ingress: []corev1.LoadBalancerIngress{
+							{
+								IP:       oldIP,
+								Hostname: oldHostname,
+							},
+						},
+					},
+				},
+			},
+			&kongv1beta1.UDPIngress{
+				Status: kongv1beta1.UDPIngressStatus{
+					LoadBalancer: corev1.LoadBalancerStatus{
+						Ingress: []corev1.LoadBalancerIngress{
+							{
+								IP:       oldIP,
+								Hostname: oldHostname,
+							},
+						},
+					},
+				},
+			},
+		}
+	}
+
+	t.Run("update not needed as no changes detected", func(t *testing.T) {
+		oldIngress := oldIngress()
+		newAddresses := []netv1.IngressLoadBalancerIngress{
+			{
+				IP:       oldIP,
+				Hostname: oldHostname,
+			},
+		}
+
+		for _, old := range oldIngress {
+			t.Run(fmt.Sprintf("%T", old), func(t *testing.T) {
+				copiedOld := old.DeepCopyObject()
+				updatedNeeded, err := UpdateLoadBalancerIngress(old, newAddresses)
+				require.NoError(t, err)
+				assert.False(t, updatedNeeded)
+				assert.Equal(t, copiedOld, old, "when update not needed, the old object shouldn't be updated")
+			})
+		}
+	})
+
+	t.Run("update needed", func(t *testing.T) {
+		oldIngress := oldIngress()
+		const (
+			newIP = "192.168.1.1"
+		)
+		newAddresses := []netv1.IngressLoadBalancerIngress{
+			{
+				IP:       newIP,
+				Hostname: oldHostname,
+			},
+		}
+
+		for _, old := range oldIngress {
+			t.Run(fmt.Sprintf("%T", old), func(t *testing.T) {
+				copiedOld := old.DeepCopyObject()
+				updatedNeeded, err := UpdateLoadBalancerIngress(old, newAddresses)
+				require.NoError(t, err)
+				assert.True(t, updatedNeeded)
+				assert.NotEqual(t, copiedOld, old, "when updated needed, the old object should be updated")
+			})
+		}
+	})
+
+	t.Run("unknown type passed should not panic", func(t *testing.T) {
```

Comment:
Nitpick: this subtest name suggest that we're actually testing not panicking which IMHO should be done by something like [`assert.NotPanics()`](https://pkg.go.dev/github.com/stretchr/testify@v1.8.1/assert#NotPanics)

Reply:
Let's add it for clarity then: [9d04f85](https://github.com/Kong/kubernetes-ingress-controller/pull/3251/commits/9d04f85b32b723f46c3ed50701556421a8593300)